chiselFrontend/src/main/scala/chisel3/core/MultiClock.scala

```diff
@@ -0,0 +1,25 @@
+// See LICENSE for license details.
+
+package chisel3.core
+
+import scala.language.experimental.macros
+
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl._
+import chisel3.internal.sourceinfo.{SourceInfo}
+
+private[chisel3] final case class ClockAndReset(clock: Clock, reset: Bool)
+
+object withClockAndReset {  // scalastyle:ignore object.name
+  def apply[T](clock: Clock, reset: Bool)(block: => T)(implicit sourceInfo: SourceInfo): T = {
```

Comment:
This PR is intended to lead in to `AsyncReset`. The idea is that `AsyncReset` will be a new type that if you pass it as the reset to `withClockAndReset` it will construct `AsyncResetRegs` (working name, suggestions welcome) and Modules where the implicit reset is an `AsyncReset`.